### PR TITLE
[TIMOB-2122] (Android) Added support for Video Recording in Android

### PR DESCRIPTION
**Pull request by m1ga**
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-2122

Reopening https://github.com/appcelerator/titanium_mobile/pull/5751 and make it work with the latest branch again.

### CODE REVIEW

**Comment by hansemannn** [comment]
Thank you @m1ga! We will take a look asap.

**Comment by ashcoding** [comment]
@m1ga Is this still in progress (not yet for review)?

**Comment by m1ga** [comment]
@ashcoding I'm still testing it on various devices. I have a problem with a stretched preview at the moment on a Moto G (2015)

**Comment by ashcoding** [comment]
@m1ga Alright. Thank you for the hard work! :+1: Let me (or @hansemannn)  know when it's ready.

**Comment by m1ga** [comment]
@hansemannn @ashcoding tested it on a HTC A9, Huawei T1-701w (tablet) and a Nexus 4. Video front/back is working fine. Fixed a last bug that onMeasure wasn't called beofre showing the first preview on the Huawei tablet.

**Comment by ashcoding** [comment]
Thank you! Will be testing it with the code provided:-
```javascript
var isRecording = false;
var isPhoto = false; // photo or video mode
 
var win = Ti.UI.createWindow({
    backgroundColor: "#fff"
});
 
var btn_cam1 = Ti.UI.createButton({
    bottom: 10,
    title: "CAM INTENT"
})
 
var btn_cam2 = Ti.UI.createButton({
    bottom: 50,
    title: "CAM OVERLAY REAR",
    id: "cam_rear"
})
var btn_cam3 = Ti.UI.createButton({
    bottom: 90,
    title: "CAM OVERLAY FRONT",
    id: "cam_front"
})
 
var btn_cam4 = Ti.UI.createButton({
    bottom: 130,
    title: "PHOTO REAR",
    id: "photo_rear"
})
var btn_cam5 = Ti.UI.createButton({
    bottom: 170,
    title: "PHOTO FRONT",
    id: "photo_front"
})
 
var btnCamera = Ti.UI.createButton({
    title: 'RECORD',
    bottom: 10
});
 
var btnCameraClose = Ti.UI.createButton({
    title: 'CLOSE',
    top: 10
});
 
var overlay = Ti.UI.createView({
    top: 0,
    right: 0,
    bottom: 0,
    left: 0
});
 
var movie = Titanium.Media.createVideoPlayer({
    movieControlStyle: Titanium.Media.VIDEO_CONTROL_EMBEDDED,
    top: 0,
    height: 200,
    width: 200,
    backgroundColor: "#000",
    autoplay:false
});
 
 
overlay.add(btnCamera);
overlay.add(btnCameraClose);
 
btnCameraClose.addEventListener('click', function(e) {
    Titanium.Media.hideCamera();
});
 
btnCamera.addEventListener('click', function(e) {
    if (isPhoto){
        // photo mode
        Titanium.Media.takePicture();
    } else {
        // video mode
        if (!isRecording) {
            Titanium.Media.startVideoCapture();
            isRecording = true;
            btnCamera.title = "STOP";
        } else {
            Titanium.Media.stopVideoCapture();
        }
    }
});
 
function onClickCam(e) {
    // intent
    Ti.Media.showCamera({
        autohide: false,
        animated: false,
        allowEditing: false,
        saveToPhotoGallery: true,
        success: function(event) {
            movie.url = event.media.nativePath;
        },
        cancel: function(event) {
            console.log("error");
        },
        error: function(error) {
            alert('error');
        },
        mediaTypes: [Titanium.Media.MEDIA_TYPE_VIDEO],
        videoMaximumDuration: 5000,
        videoQuality: Titanium.Media.QUALITY_HIGH
    });
}
 
 
function onClickCam2(e) {
    // native camera
    var which = Ti.Media.CAMERA_REAR;
    var type = Titanium.Media.MEDIA_TYPE_VIDEO;
    var autohide = false;
    isPhoto = false;
    
    if (e.source.id == "cam_front") {
        which = Ti.Media.CAMERA_FRONT;
    }
    if (e.source.id=="photo_front") {
        which = Ti.Media.CAMERA_FRONT;
        type = Ti.Media.MEDIA_TYPE_PHOTO;
        isPhoto = true;
        autohide = true;
    }
    if (e.source.id=="photo_rear") {
        type = Ti.Media.MEDIA_TYPE_PHOTO;
        isPhoto = true;
        autohide = true;
    }
    
    console.log(type + " " + which);
    
    Ti.Media.showCamera({
        overlay: overlay,
        autohide: autohide,
        animated: false,
        allowEditing: false,
        saveToPhotoGallery: true,
        whichCamera: which,
        success: function(event) {
            if (!isPhoto) movie.url = event.media.nativePath;
            isRecording = false;
            btnCamera.title = "RECORD";
        },
        cancel: function(event) {
            console.log("error");
        },
        error: function(error) {
            alert('error');
        },
        mediaTypes: [type],
        videoMaximumDuration: 5000,
        videoQuality: Titanium.Media.QUALITY_HIGH
    });
}
 
btn_cam1.addEventListener("click", onClickCam);
btn_cam2.addEventListener("click", onClickCam2);
btn_cam3.addEventListener("click", onClickCam2);
btn_cam4.addEventListener("click", onClickCam2);
btn_cam5.addEventListener("click", onClickCam2);
 
win.add(btn_cam1);
win.add(btn_cam2);
win.add(btn_cam3);
win.add(btn_cam4);
win.add(btn_cam5);
win.add(movie);
win.open();
```

**Comment by m1ga** [comment]
Added a small fix for a Samsung device that didn't return a videosize.
From the Android doc it can return null if preview == output:

`a list of Size object if camera has separate preview and video output; otherwise, null is returned.`

Tested successfully on an old Samsung Galaxy S2

**Comment by ashcoding** [comment]
Awesome work! Is it ready for review and testing?

**Comment by m1ga** [comment]
Yes,ready for testing.

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java [hunk 1]
```diff
@@ -47,22 +49,30 @@
...
+	private static final String MEDIA_TYPE_PHOTO = "public.image";
+	private static final String MEDIA_TYPE_VIDEO = "public.video";
+	
 	private TiViewProxy localOverlayProxy = null;
 	private SurfaceView preview;
 	private PreviewLayout previewLayout;
 	private FrameLayout cameraLayout;
 	private boolean previewRunning = false;
-	private int currentRotation;
+	private static int currentRotation;
```
Comment (ashcoding) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java]: Just want to check, how does making it `static` help?

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java [hunk 2]
```diff
@@ -102,11 +120,16 @@ public void run()
 		@Override
 		protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec)
 		{
+			final int width = resolveSize(getSuggestedMinimumWidth(), widthMeasureSpec);
+        	final int height = resolveSize(getSuggestedMinimumHeight(), heightMeasureSpec);
```
Comment (ashcoding) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java]: Slight alignment issue on lines 124 and 125.

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java [hunk 3]
```diff
@@ -377,6 +425,143 @@ private void stopPreview()
...
+	        profile.videoFrameHeight = optimalVideoSize.height;
+		} else {
+			Size videoSize = getOptimalPictureSize(supportedVideoSizes);
+			if (videoSize!=null){
+				profile.videoFrameWidth = videoSize.width;
+		        profile.videoFrameHeight = videoSize.height;
+			}
+		}
+		int result  = TiApplication.getInstance().getRootActivity().checkCallingOrSelfPermission("android.permission.RECORD_AUDIO");
+    	if (result == PackageManager.PERMISSION_GRANTED) {
```
Comment (ashcoding) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java]: Alignment/tab is wrong for line 466.

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java [hunk 4]
```diff
@@ -398,29 +583,37 @@ public void finish()
...
-		// Try to find an size match aspect ratio and size
-		for (Size size : sizes) {
-			double ratio = (double) size.width / size.height;
-			if (Math.abs(ratio - targetRatio) < minAspectDiff) {
-				optimalSize = size;
-				minAspectDiff = Math.abs(ratio - targetRatio);
-			}
-		}
+	{		
+		final double ASPECT_TOLERANCE = 0.1;
```
Comment (ashcoding) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java]: Alignment/Tab is wrong at line 587.

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java [hunk 5]
```diff
@@ -602,7 +799,7 @@ private void checkWhichCameraAsDefault(){
 		if (backCameraId == Integer.MIN_VALUE && frontCameraId != Integer.MIN_VALUE) {
 			TiCameraActivity.whichCamera = MediaModule.CAMERA_FRONT;
 		} else {
-			TiCameraActivity.whichCamera = MediaModule.CAMERA_REAR;
+			//TiCameraActivity.whichCamera = MediaModule.CAMERA_REAR;
```
Comment (ashcoding) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java]: Why is this removed?

Comment (m1ga) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java]: Because that would always set it to the rear camera even if you want to start it with the front camera. It would override the whichCamera parameter in openCamera

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java (hunk 1 above)
Comment (m1ga) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java]: you're right. I had a static function before and the compiler complaint that I'm using a non static variable. The function got removed so it doesn't need to be static anymore.

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java (hunk 5 above)
Comment (ashcoding) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java]: I see and understand. In that case, please remove/comment out the else bracket as well if you are removing that portion of the logic.

Comment (m1ga) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java]: Just pushed an update

File: android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java [hunk 6]
```diff
@@ -602,7 +798,7 @@ private void checkWhichCameraAsDefault(){
 		if (backCameraId == Integer.MIN_VALUE && frontCameraId != Integer.MIN_VALUE) {
 			TiCameraActivity.whichCamera = MediaModule.CAMERA_FRONT;
 		} else {
```
Comment (ashcoding) [inline android/modules/media/src/java/ti/modules/titanium/media/TiCameraActivity.java]: Could you remove ```} else {``` as well?

**Comment by ashcoding** [comment]
I'm also trying to get audio in the video but currently not able to do so.
I added ```<uses-permission android:name="android.permission.RECORD_AUDIO"/>``` in the manifest. Anything else I need to do based on the sample app I'm using above?

**Comment by ashcoding** [comment]
Code reviewed and functionally tested. Working.
**Accepted**
Really great work here!

I'll be doing just one PR later to fix minor formatting issues.

**Comment by ashcoding** [comment]
**Merging**

**Comment by ashcoding** [comment]
Fixing formatting here: https://github.com/appcelerator/titanium_mobile/pull/7965

**Comment by m1ga** [comment]
Thanks!

**Comment by hansemannn** [comment]
@m1ga There seems to be an issue with this PR: https://jira.appcelerator.org/browse/AC-5767
